text/0000-hash-simplification.md

```diff
@@ -0,0 +1,405 @@
+   algorithm is seeded with a 128-bit key.
+
+These two measures ensure that each `HashMap` is randomly ordered, even if the
+same keys are inserted in the same order. As a result, it is quite difficult to
+mount a DoS attack against a `HashMap` as it is difficult to predict what
+collisions will happen.
+
+The `Hash` trait proposed above, however, does not allow SipHash to be
+implemented generally any more. For example `#[derive(Hash)]` will no longer
+leverage SipHash. Additionally, there is no input of state into the `hash`
+function, so there is no random state per-`HashMap` to generate different hashes
+with.
+
+Denial of service attacks against hash maps are no new phenomenon, they are
+[well](http://www.ocert.org/advisories/ocert-2011-003.html)
+[known](http://lwn.net/Articles/474912/)
+and have been reported in
+[Python](http://bugs.python.org/issue13703),
+[Ruby](https://www.ruby-lang.org/en/news/2011/12/28/denial-of-service-attack-was-found-for-rubys-hash-algorithm-cve-2011-4815/)
+([other ruby](https://www.ruby-lang.org/en/news/2012/11/09/ruby19-hashdos-cve-2012-5371/)),
+[Perl](http://blog.booking.com/hardening-perls-hash-function.html),
+and many other languages/frameworks. Rust has taken a fairly proactive step from
+the start by using a strong and randomly seeded algorithm since `HashMap`'s
+inception.
+
+In general the standard library does not provide many security-related
+guarantees beyond memory safety. For example the new `Read::read_to_end`
+function passes a safe buffer of uninitialized data to implementations of
+`read` using various techniques to prevent memory safety issues. A DoS attack
+against a hash map is such a common and well known exploit, however, that this
+RFC considers it critical to consider the design of `Hash` and its relationship
+with `HashMap`.
+
+## Mitigation of DoS attacks
+
+Other languages have mitigated DoS attacks via a few measures:
+
+* [C++ specifies][cpp-hash] that the return value of `hash` is not guaranteed to
+  be stable across program executions, allowing for a global salt to be mixed
+  into hashes calculated.
+* [Ruby has a global seed][ruby-seed] which is randomly initialized on startup
+  and is used when hashing blocks of memory (e.g. strings).
+* PHP and Tomcat have added limits to the maximum amount of keys allowed from a
+  POST HTTP request (to limit the size of auto-generated maps). This strategy is
+  not necessarily applicable to the standard library.
+
+[cpp-hash]: http://en.cppreference.com/w/cpp/utility/hash
+[ruby-seed]: https://github.com/ruby/ruby/blob/193ad64359b8ebcd77a2cba50a62d64311e26b22/random.c#L1248-L1251
+
+It [has been claimed](http://bugs.python.org/issue13703#msg150558), however,
+that a global seed may only mitigate some of the simplest attacks. The primary
+downside is that a long-running process may leak the "global seed" through some
+other form which could compromise maps in that specific process.
```

Comment:
The approach recommended by the SipHash paper is to obtain a random key from the OS at boot time and use that for all subsequent hashes. Reseeding/rekeying isn't discussed and shouldn't be necessary.